pytorch_lightning/metrics/classification/utils.py

```diff
@@ -0,0 +1,441 @@
+            "You have set `top_k` above 1, but your data is not (multi-dimensional) multi-class"
+            " with probability predictions."
+        )
+    if is_multiclass is False:
+        raise ValueError("If you set `is_multiclass=False`, you can not set `top_k`.")
+    if top_k >= implied_classes:
+        raise ValueError("The `top_k` has to be strictly smaller than the `C` dimension of `preds`.")
+
+
+def _check_classification_inputs(
+    preds: torch.Tensor,
+    target: torch.Tensor,
+    threshold: float,
+    num_classes: Optional[int],
+    is_multiclass: bool,
+    top_k: Optional[int],
+) -> str:
+    """Performs error checking on inputs for classification.
+
+    This ensures that preds and target take one of the shape/type combinations that are
+    specified in ``_input_format_classification`` docstring. It also checks the cases of
+    over-rides with ``is_multiclass`` by checking (for multi-class and multi-dim multi-class
+    cases) that there are only up to 2 distinct labels.
+
+    In case where preds are floats (probabilities), it is checked whether they are in [0,1] interval.
+
+    When ``num_classes`` is given, it is checked that it is consitent with input cases (binary,
+    multi-label, ...), and that, if availible, the implied number of classes in the ``C``
+    dimension is consistent with it (as well as that max label in target is smaller than it).
+
+    When ``num_classes`` is not specified in these cases, consistency of the highest target
+    value against ``C`` dimension is checked for (multi-dimensional) multi-class cases.
+
+    If ``top_k`` is set (not None) for inputs which are not (multi-dimensional) multi class
+    with probabilities, then an error is raised. Similarly if ``top_k`` is set to a number
+    that is higher than or equal to the ``C`` dimension of ``preds``.
+
+    Preds and target tensors are expected to be squeezed already - all dimensions should be
+    greater than 1, except perhaps the first one (N).
+
+    Args:
+        preds: Tensor with predictions (labels or probabilities)
+        target: Tensor with ground truth labels, always integers (labels)
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
+        num_classes:
+            Number of classes. If not explicitly set, the number of classes will be infered
+            either from the shape of inputs, or the maximum label in the ``target`` and ``preds``
+            tensor, where applicable.
+        top_k:
+            Number of highest probability entries for each sample to convert to 1s - relevant
+            only for (multi-dimensional) multi-class inputs with probability predictions. The
+            default value (``None``) will be interepreted as 1 for these inputs.
+
+            Should be left unset (``None``) for all other types of inputs.
+        is_multiclass:
```

Comment:
I am starting to understand the meaning of this parameter, and I am willing to include it for now to see how it plays out when actually integrated with some of our metrics. However, I really in doubt if `is_multiclass` is the correct name for it. If is see the argument name (without reading the documentation) I would suspect that I need to set it to `True` if I am running any form of multiclass classification

Reply:
Ok great :) What would you suggest as a name?

Reply:
Also, I prepared an essay about the use of `is_multiclass`, to better explain what it is meant for and why I think it is better than alternatives:

## How `is_multiclass` works

First, let me clear up some confusion about it: this is not something that needs to be set (i.e. changed from default `None`) in ~99% of cases. If the inputs fall into any of these cases (listed in the new table in documentation):
- binary (with probability predictions)
- (multi-dim) multi-class (with label or probability predictions)
- multi-label (with probability predictions)

the parameter *does not* need to be set. The new formatting function is "intelligent" enough to correctly deduce the type and apply proper formatting in these cases. I have updated the description of the parameter in the docstring to reflect this and avoid confusing users who don't need to use this parameter.

There are only 2 special cases where this parameter needs to be used. They concern cases where inputs appear as one case, but are actually another - and in a way that is ambiguous enough to make determining the actual case impossible without explicit user instruction, or where for some reason the user wants to treat inputs as a different case from what they appear. These cases are:

- **Use case 1**: for multi-class or multi-dim multi-class with 2 classes, if the user wants to treat inputs as binary or multi-label, respectively. To be explicit, here's what I mean

    ``` python
    # Appears as multi-class, is actually binary
    preds = [0,1,1,0,1]

    # Appears as multi-dim multi-class, is actually multi-label
    preds = [[0,1,0], [1,0,1], [0,0,1]]
    ```

    These 2 cases are currently (i.e. in master, before my PRs) included as test cases for metrics, so I assume that they are considered important enough to be supported.

    It's clear why the type can not be (correctly) automatically infered here: even if we assume that multi-class data with 2 classes does not exist in reality, it can be the case that this you have, for example, 3 classes in your data, but it just happens that only 2 were present in a batch.

- **Use case 2**: for binary or multi-label data, if the user wants to treat it as multi-class or multi-dim multi-class with 2 classes. This would be the case if the model gives prediction only for one class, but the user would like to compute the metrics for the implied 0, as well as the "default" 1 class. Here's what I mean (one example of binary data)

    ``` python
    preds = [0.1, 0.7, 0.2, 0.6]

    # without setting is_multiclass
    preds_transformed = [[0], [1], [0], [1]]

    # setting is_multiclass=True
    preds_transformed = [[1, 0], [0, 1], [1, 0], [1, 0]]
    ```

    An example of this would be image segmentation: if the prediction is binary (e.g. tissue is cancerous/not cancerous), usually only one class (foreground) is predicted. But the user might want to see the metrics for the "background" predictions as well. Sure, the model could be changed to output two values, the user could transform the outputs so that they contain 2 classes before feeding them into the metrics, etc. - but this all feels like extra "engineering code" that PL is supposed to do away with.

## Alternatives

Most importantly, here's why I chose to use ``is_multiclass`` and not the alternatives.

* **Alternative 1**: just use `num_classes`: This version would imply that the user wants to treat data a binary/multi-label if `num_classes=1`, and muli-class if `num_classes > 1`. There are a couple things I don't like about this alternative:

    - Some metrics (like recall, precision...), use the `num_classes` argument to determine the shape of the state variables. So setting `num_classes=1` to signal that data is actually multi-label (see the first use-case in the previous section), would cause the metrics to error out on the update step. To avoid these such metrics would then need to introduce another parameter (`num_classes_in_data`?). I think this alone disqualifies this alternative, but there are more reasons :)

    - This prevents `num_classes` from being used in data validation in the cases where it is used to override the aparent type. This then pushes errors down to the update step, where the error message (which will result, for example, from trying to add two tensors of incompatible sizes) will be much less clear to the user.

    - This prevents solving Use case 2 above: If multi-label data consists of 2 classes (`C` dimension equals 2), then setting `num_classes=2` can't specify whether the user wants to treat this as multi-dim multi-class data, or is just saying that this is multi-label data with 2 classes.

* **Alternative 2**: just use `num_classes` + abandon support for binary/multi-label predictions with labels. This will require users to manually cast these predictions to `float` before feeding them into the metrics. The prolems with this are

    - As mentioned before, binary/multi-label predictions with labels inputs are currently among the test cases, so I assume that it had been decided they must be supported.

    - Same as for alternative 1, Use case 2 can not be supported.


I think the only way we could possibly avoid using ``is_multiclass`` parameter is by requiring binary/multi-label inputs to be floats, and requiring users to transform single-class data to 2-class manually (for Use Case 2). But I think that overall this is not a good tradeoff.

Reply:
Quick question about 
```
preds = [0.1, 0.7, 0.2, 0.6]

# without setting is_multiclass
preds_transformed = [[0], [1], [0], [1]]

# setting is_multiclass=True
preds_transformed = [[1, 0], [0, 1], [1, 0], [1, 0]]
```
Should it be 

```
preds_transformed = [[0, 1], [1, 0], [0, 1], [1, 0]] ?
```

Reply:
Is should actually be
```
preds_transformed = [[1, 0], [0, 1], [1, 0], [0, 1]]
```
If `prob>=0.5`, then this is class `1`, so `[0,1]`, if `prob < 0.5` it is class `0`, so `[1, 0]`.